gtk/gtkpagesetup.c

```diff
@@ -837,3 +837,112 @@ gtk_page_setup_to_key_file (GtkPageSetup *setup,
+  ppd_name = gtk_paper_size_get_ppd_name (paper_size);
+  display_name = gtk_paper_size_get_display_name (paper_size);
+
+  if (ppd_name != NULL)
+    g_variant_builder_add (&builder, "{sv}", "PPDName", g_variant_new_string (ppd_name));
+  else
+    g_variant_builder_add (&builder, "{sv}", "Name", g_variant_new_string (name));
+
+  if (display_name != NULL)
+    g_variant_builder_add (&builder, "{sv}", "DisplayName", g_variant_new_string (display_name));
+
+  g_variant_builder_add (&builder, "{sv}", "Width", g_variant_new_double (gtk_paper_size_get_width (paper_size, GTK_UNIT_MM)));
+  g_variant_builder_add (&builder, "{sv}", "Height", g_variant_new_double (gtk_paper_size_get_height (paper_size, GTK_UNIT_MM)));
+
+  g_variant_builder_add (&builder, "{sv}", "MarginTop", g_variant_new_double (gtk_page_setup_get_top_margin (setup, GTK_UNIT_MM)));
+  g_variant_builder_add (&builder, "{sv}", "MarginBottom", g_variant_new_double (gtk_page_setup_get_bottom_margin (setup, GTK_UNIT_MM)));
+  g_variant_builder_add (&builder, "{sv}", "MarginLeft", g_variant_new_double (gtk_page_setup_get_left_margin (setup, GTK_UNIT_MM)));
+  g_variant_builder_add (&builder, "{sv}", "MarginRight", g_variant_new_double (gtk_page_setup_get_right_margin (setup, GTK_UNIT_MM)));
+
+  orientation = enum_to_string (GTK_TYPE_PAGE_ORIENTATION,
+                                gtk_page_setup_get_orientation (setup));
+  g_variant_builder_add (&builder, "{sv}", "Orientation", g_variant_new_string (orientation));
+  g_free (orientation);
+
+  return g_variant_builder_end (&builder);
+}
+
+/**
+ * gtk_page_setup_new_from_gvariant:
+ * @variant: an a{sv} #GVariant
+ *
+ * Desrialize a page setup from an a{sv} variant in
+ * the format produced by gtk_page_setup_to_gvariant().
+ *
+ * Returns: (transfer full): a new #GtkPageSetup object
+ *
+ * Since: 3.22
+ */
+GtkPageSetup *
+gtk_page_setup_new_from_gvariant (GVariant *variant)
+{
+  GtkPageSetup *setup;
+  const char *orientation;
+  gdouble margin;
+  gdouble size;
+  GKeyFile *key_file;
+  const char *name;
+  GtkPaperSize *paper_size;
+
+  g_return_val_if_fail (g_variant_is_of_type (variant, G_VARIANT_TYPE_VARDICT), NULL);
+
+  setup = gtk_page_setup_new ();
+
+  key_file = g_key_file_new ();
+  if (g_variant_lookup (variant, "PPDName", "&s", &name))
+    g_key_file_set_string (key_file, "Page Setup", "PPDName", name);
+  else if (g_variant_lookup (variant, "Name", "&s", &name))
+    g_key_file_set_string (key_file, "Page Setup", "Name", name);
+  if (g_variant_lookup (variant, "DisplayName", "&s", &name))
+    g_key_file_set_string (key_file, "Page Setup", "DisplayName", name);
+  if (g_variant_lookup (variant, "Width", "d",  &size))
+    g_key_file_set_double (key_file, "Page Setup", "Width", size);
+  if (g_variant_lookup (variant, "Height", "d",  &size))
+    g_key_file_set_double (key_file, "Page Setup", "Height", size);
+  paper_size = gtk_paper_size_new_from_key_file (key_file, "Page Setup", NULL);
```

Comment:
It feels a bit awkward to go through variant -> key file -> struct here, but perhaps it's OK for the time being.